pkg/doc/verifiable/credential.go

```diff
@@ -452,90 +436,127 @@ func decodeCredentialSchema(data []byte) ([]TypedID, error) {
+// For JSON bytes input, the output marshalled JSON is the same value.
+// For serialized JWT input, the output is the result of decoding `vc` claim from JWT.
+// The output Credential and marshalled JSON can be used for extensions of the base data model
+// by checking ExtraFields of Credential and/or unmarshalling the JSON to custom date structure.
+func NewCredential(vcData []byte, opts ...CredentialOpt) (*Credential, []byte, error) {
+	// Apply options.
+	crOpts := parseCredentialOpts(opts)
 
+	// Decode credential (e.g. from JWT).
 	vcDataDecoded, err := decodeRaw(vcData, crOpts)
 	if err != nil {
-		return nil, err
+		return nil, nil, fmt.Errorf("decode new credential: %w", err)
 	}
 
-	// unmarshal VC from JSON
+	// Unmarshal raw credential from JSON.
 	var raw rawCredential
-
 	err = json.Unmarshal(vcDataDecoded, &raw)
+
 	if err != nil {
-		return nil, fmt.Errorf("new credential: %w", err)
+		return nil, nil, fmt.Errorf("unmarshal new credential: %w", err)
 	}
 
-	schemas, err := loadCredentialSchemas(&raw, vcDataDecoded)
-	if err != nil {
-		return nil, err
+	// Load custom credential schemas if defined.
+	var schemas []TypedID
+	if raw.Schema != nil {
+		schemas, err = loadCredentialSchemas(vcDataDecoded)
+		if err != nil {
+			return nil, nil, fmt.Errorf("load schemas of new credential: %w", err)
+		}
+	} else {
+		schemas = make([]TypedID, 0)
 	}
 
+	// Validate raw credential.
 	err = validate(vcDataDecoded, schemas, crOpts)
 	if err != nil {
-		return nil, err
+		return nil, nil, fmt.Errorf("validate new credential: %w", err)
 	}
 
-	cred := crOpts.template()
-
-	err = cred.fill(&raw)
+	// Create credential from raw.
+	vc, err := newCredential(&raw, schemas)
 	if err != nil {
-		return nil, err
+		return nil, nil, fmt.Errorf("build new credential: %w", err)
 	}
 
-	cred.Schemas = schemas
+	return vc, vcDataDecoded, nil
+}
 
-	for _, decoder := range crOpts.decoders {
-		err = decoder(vcDataDecoded, cred)
-		if err != nil {
-			return nil, err
+// CustomCredentialProducer is a factory for Credentials with extended data model.
+type CustomCredentialProducer interface {
+	// Accept checks if producer is capable of building extended Credential data model.
+	Accept(vc *Credential) bool
+
+	// Apply creates custom credential using base credential and its JSON bytes.
+	Apply(vc *Credential, dataJSON []byte) (interface{}, error)
+}
+
+// CreateCustomCredential creates custom extended credentials from bytes which could be marshalled JSON
+// or serialized JWT. It decodes input bytes to the base Verifiable Credential using NewCredential().
+// It then checks all producers to find the appropriate which is capable of building extended Credential data model.
+// If none of producers accept the credential, the base credential is returned.
+func CreateCustomCredential(
+	vcData []byte,
+	producers []CustomCredentialProducer,
+	opts ...CredentialOpt) (interface{}, error) {
+	vcBase, vcBytes, credErr := NewCredential(vcData, opts...)
```

Comment:
I guess you are thinking the second return value is useful beyond its usage in CreateCustomCredential?

Reply:
I think - yes, because in case of VC data model extension we can either use `ExtraFields` of `verifiable.Credential` or make unmarshaling of bytes to a custom structure that defines extra fields. An example can be found [here](https://github.com/kdimak/aries-framework-go/blob/5fa6c7f7402094770ac7597021b1da58f26e86b1/pkg/doc/verifiable/credential_extension_test.go#L40).
But if I'm alone in this opinion, I can revert and return just `Credential`  (without bytes).

Reply:
I think we should have another TODO to rename ExtraFields.